docs/api-spec.md

```diff
@@ -48,269 +72,447 @@ There is no formal specification of the Kubernetes API and Resource Model. This
-| `name`               | string             | REQUIRED    |
-| `labels`             | map<string,string> | REQUIRED    |
-| `annotations`        | map<string,string> | REQUIRED    |
-| `creationTimestamp`* | string             | REQUIRED    |
-| `uid`                | string             | RECOMMENDED |
-| `resourceVersion`    | string             | RECOMMENDED |
-| `generation`         | int64              | RECOMMENDED |
-| `namespace`          | string             | RECOMMENDED |
-| `generateName`**     | string             | RECOMMENDED |
+| Field         | Type                                              | Requirement | Notes |
+| ------------- | ------------------------------------------------- | ----------- | ---- |
+| `description` | string                                            | RECOMMENDED |      |
+| `params`      | [][`ParamSpec`](#paramspec)                       | REQUIRED    |      |
+| `steps`       | [][`Step`](#step)                                 | REQUIRED    |      |
+| `sidecars`    | [][`Sidecar`](#sidecar)                           | RECOMMENDED |      |
+| `results`     | [][`TaskResult`](#taskresult)                     | REQUIRED    |      |
+| `workspaces`  | [][`WorkspaceDeclaration`](#workspacedeclaration) | REQUIRED    |      |
 
-\* `creationTimestamp` MUST be populated by the implementation, in [RFC3339](https://tools.ietf.org/html/rfc3339).
+### ParamSpec
 
-** If `generateName` is supported by the implementation, when it is specified at creation, it MUST be prepended to a random string and set as the `name`, and not set on the subsequent response.
+Declares a parameter whose value has to be provided at runtime
 
-#### Spec
+| Field Name    | Field Type                  | Requirement         | Notes                                                                                                                                                                                    |
+| ------------- | --------------------------- | ------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `name`        | string                      | REQUIRED            |                                                                                                                                                                                          |
+| `description` | string                      | REQUIRED            |                                                                                                                                                                                          |
+| `type`        | [`ParamType`](#paramtype)   | REQUIRED (see note) | The values `string` and `array` for this field are REQUIRED, and the value `object` is RECOMMENDED.                                                                                      |
```

Comment:
Why is "object" RECOMMENDED and "array" REQUIRED?

Reply:
Oh, I guess they are not stable yet?

Reply:
yup, array has been around for much longer and is stable. object is not yet stable